python/cuml/common/doc_utils.py

```diff
@@ -0,0 +1,283 @@
+                Acceptable dense formats: CUDA array interface compliant objects like \
+                CuPy, cuDF DataFrame/Series, NumPy ndarray and Pandas \
+                DataFrame/Series.',  # noqa
+
+    'return_sparse':
+    'return_sparse : bool, optional (default = {}) \n \
+            Ignored when the model is not fit on a sparse matrix \
+            If True, the method will convert the result to a \
+            cupy.sparse.csr_matrix object. \n \
+            NOTE: Currently, there is a loss of information when converting \
+            to csr matrix (cusolver bug). Default will be switched to True \
+            once this is solved.',
+
+    'sparse_tol':
+    'sparse_tol : float, optional (default = {}) \n\
+            Ignored when return_sparse=False. \
+            If True, values in the inverse transform below this parameter \
+            are clipped to 0.',
+
+    '_custom_docstring_default':
+    '{}: {} (default = {}) \n \
+        {}',
+
+    '_custom_docstring_no_default':
+    '{}: {} \n \
+        {}'
+}
+
+_parameter_possible_values = ['name',
+                              'type',
+                              'shape',
+                              'default',
+                              'description',
+                              'accepted']
+
+_return_values_docstrings = {
+    'dense':
+    '{} : cuDF, CuPy or NumPy object depending on cuML\'s output type configuration, shape = {}\n \
+        {} \n For more information on how to configure cuML\'s output type, \
+        refer to: `Output Data Type Configuration`_.',  # noqa
+
+    'dense_sparse':
+    '{} : cuDF, CuPy or NumPy object depending on cuML\'s output type configuration, cupy.sparse for sparse output, shape = {}\n \
+        {} \n For more information on how to configure cuML\'s dense output type, \
+        refer to: `Output Data Type Configuration`_.',  # noqa
+
+    'dense_datatype':
+    'cuDF, CuPy or NumPy object depending on cuML\'s output type \
+    configuration, shape ={}',
+
+    'dense_sparse_datatype':
+    'cuDF, CuPy or NumPy object depending on cuML\'s output type \
+    configuration, shape ={}',
+
+    'custom_type':
+    '{} : {} \n \
+        {}'
+
+}
+
+_return_values_possible_values = ['name',
+                                  'type',
+                                  'shape',
+                                  'description']
+
+_simple_params = ['return_sparse',
+                  'sparse_tol',
+                  'sample_weight']
+
+
+def generate_docstring(X='dense',
+                       X_shape='(n_samples, n_features)',
+                       y='dense',
+                       y_shape='(n_samples, 1)',
+                       convert_dtype_cast=False,
+                       skip_parameters=[],
+                       skip_parameters_heading=False,
+                       parameters=False,
+                       return_values=False):
+    def deco(func):
+        @wraps(func)
```

Comment:
phew, this function got a little complex... maybe there is room to break it up?

Reply:
I added comments to make it more clear. Didn't want to break it into further functions to avoid any potential overheads (even if small)